contributors/design-proposals/scheduling/gang-scheduling.md

```diff
@@ -0,0 +1,203 @@
+
+1. Existing workload can use this feature without (or with a few) configuration changes
+2. Pods of a group/gang may have different `PodSpec` (and/or belong to different collections)
+3. Existing controllers which are responsible for managing life cycle of collections work well with this feature
+
+To meet the requirements above, the following **Kind** is introduced in core part under `scheduling.k8s.io/v1alpha1` **Group**/**Version**.
+
+```go
+// PodGroup defines the scheduling requirement of a pod group
+type PodGroup struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+	// Spec defines the behavior of a pod group.
+	// +optional
+    Spec PodGroupSpec
+
+	// Status represents the current information about a pod group.
+	// This data may not be up to date.
+	// +optional
+    Status PodGroupStatus
+}
+
+// LifecyclePolicy represents the lifecycle policy of PodGroup.
+type LifeCyclePolicy struct {
+    // The action that will be taken to the PodGroup according to Event.
+    // One of "Restart", "None".
+    // Default to None.
+    // +optional
+    Action Action
+
+    // The Event recorded by scheduler; the controller takes actions
+    // according to this Event.
+    // One of "PodFailed", "Unschedulable".
+    // +optional
+    Event Event
+
+    // Timeout is the grace period for controller to take actions.
+    // Default to nil (take action immediately).
+    // +optional
+    Timeout *metav1.Duration
+}
+
+// PodGroupSpec represents the template of a pod group.
+type PodGroupSpec struct {
+    // NumMember defines the number of members/tasks to run the pod group;
+    // if there's not enough resources to start all tasks, the scheduler
+    // will not start anyone.
+    NumMember int
+
+    // Policy defines the policy of PodGroup lifecycle.
+    // Default to 'Action: None, PodPhase: Failed'
+    // +optional
+    Policy []LifeCyclePolicy
+}
+
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+    // The number of actively running pods.
+    // +optional
+    Running int32
+
+    // The number of pods which reached phase Succeeded.
+    // +optional
+    Succeeded int32
+
+    // The number of pods which reached phase Failed.
+    // +optional
+    Failed int32
+}
+```
+
+The `PodGroup`, which is a namespaced object, specifies the behavior and status of a pod group, e.g. number of pods in a group, lifecycle policy. To define which pods are member of `PodGroup`, the following annotation key is introduced for `Pod`.
+
+```go
+scheduling.k8s.io/group-name
+```
+
+The `scheduling.k8s.io/group-name` annotation specifies the `PodGroup` that it belongs to; and the pod can only belong to the `PodGroup` in the same namespace. The pod, controlled by different collections, can also belong to the same `PodGroup`.  Because of performance concern,  it does not use `LabelSelector` to build the relationship between `PodGroup` and `Pod`.
```

Comment:
Perhaps I'm missing something, but how does this support pods with different specs in the same group? Specifically, suppose I have 2 ReplicaSets, "foo" and "bar". I want foo pods & bar pods to be scheduled together, in a PodGroup. Is there any way to guarantee that the group will consist of a foo & bar pod, and not 2 foos, or 2 bars?

Reply:
> but how does this support pods with different specs in the same group? 

By other algorithm plugins to work together with gang-scheduling.

> Is there any way to guarantee that the group will consist of a foo & bar pod, and not 2 foos, or 2 bars?

What's case of that? Currently, a job may include master and worker tasks; so there's TaskPriority plugin in kube-arbitrator to make sure master tasks started firstly in PodGroup.

Reply:
@tallclair Gang scheduling addresses the problem of scheduling a number of pods together. For example, you may want to schedule all members of two ReplicaSets together. Gang scheduling is not addressing pod affinity or pod dependency issues.
In the case of running all pods in multiple collections together, one can provide the name of one PodGroup in their spec/annotation. The PodGroup.Spec.NumMembers is then set to the sum of number of pods in the two collections. This configuration guarantees that they will be scheduled together.